[2.3] Static Code Analysis for Components

kalessil opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

Static Code Analysis with Php Inspections (EA Extended), no functional changes:
- resolved possible PHP Fatal in \Symfony\Component\BrowserKit\Cookie::__toString
- resolved callable name case mismatches

Tobion commented on src/Symfony/Component/Translation/MessageCatalogue.php:
```diff
@@ -232,7 +232,7 @@ public function getResources()
     public function addResource(ResourceInterface $resource)
     {
-        $this->resources[$resource->__toString()] = $resource;
+        $this->resources[(string) $resource] = $resource;
```
what is the point of that?

kalessil replied:
Point is not to use magic methods directly (as magic methods are PHP internals).

kalessil replied:
As more speaking examples (not changed by PR. so remains in components):
```php
        $object = new Magician();
        $object->__set('magicProperty', 'foobar');

        //or

        $this->assertEquals('foobar', $object->__get('magicProperty'));
```

nicolas-grekas replied:
There is a downside: casting triggers a memcopy, but calling doesn't

kalessil replied:
Good argument against. What is generally agreed approach for components? 

I asking because it's something like just 20 places where __toString is used.

kalessil replied:
And just 2 of them not related to tests.

nicolas-grekas replied:
There is no general approach. I'd say: use the cast when unsure about the type, the call when sure about type+method.

kalessil replied:
Sounds reasonable, I'll revert string castings introduced by this PR.

Tobion commented on src/Symfony/Component/BrowserKit/Cookie.php:
```diff
@@ -92,12 +92,11 @@ public function __toString()
 
-            if ($dateTime === false) {
-                // this throw will provoke PHP fatal
-                throw new \UnexpectedValueException(sprintf('The cookie expiration time "%s" is not valid.'), $this->expires);
```
since __toString cannot throw an exception, the validation of `$this->expires` should be done in the setter, i.e. the constructor in this case.

Tobion replied:
also the phpdoc of the method seems wrong: `@param string $expires`. I think it must be `int|null`. Same for the getter.

kalessil replied:
Will adjust. These changes will not breaking BC, right?

I asking because validation can trigger an exception.

nicolas-grekas replied:
fixing a fatal error is not a BC break :)

fabpot replied:
@kalessil Can you make the change? Moving the validation into the constructor and throw the exception there?

kalessil replied:
@fabpot : I'll do as proposed, just need time for this (can promise to finish untill end of May).

kalessil replied:
@fabpot : moved validation to constructor.

kalessil committed:
Php Inspections (EA Extended):
    - resolved possible PHP Fatal in \Symfony\Component\BrowserKit\Cookie::__toString
    -resolved implicit magic methods calls
    -resolved callable name case mismatches

nicolas-grekas commented:
:+1:

fabpot commented:
Thank you @kalessil.
